### ISSUE TITLE
Support for a new "LanguageType" Autocommand event

### ISSUE THREAD

**Issue by stevenproctor** [neovim/neovim#20562]
### Feature already in Vim?

No

### Feature description

With TreeSitter and the ability to have injections, NeoVim can recognize different languages inside the same file.

Is it feasible to get an Autocommand event for a TreeSitter injection/language change based on the cursor's position?

e.g., My file is HTML, and I just entered a JavaScript block. My language is Rust, and I am in embedded SQL.  I am editing a language in a fenced code block in Markdown.

**Comment by clason** [comment]
No, as tree-sitter has nothing to do with the cursor position. You'd handle this the other way around: with a `CursorHold` autocommand that gets the node under the cursor (there's a function for that) and checks its language.

**Comment by zeertzjq** [comment]
I think this can be implemented using `CursorMoved`.

**Comment by stevenproctor** [comment]
Yeah, I realized that I could hook things into the Cursor events myself.

I was thinking about a LanguageType event because so many plugins use FileType already for language specific behavior, even if it is not operation against the file itself.

It was the idea that there might be a more general Language based  Event hook that might make sense for plugins/code to hook into even if it is not triggered by Neovim in any other way than being triggered by FileType.

Looking through the documentation is there a way to publish custom auto command events myself if I do the checks based off CursorMoved or CursorHold?

**Comment by glepnir** [comment]
like `vim.api.nvim_exec_autocmds('LanguageType', opts)` usage is

```lua
vim.api.nvim_create_autocmd('LanguageType', {
  callback = function()
  end
})
```

**Comment by zeertzjq** [comment]
Should be a `User` event. You cannot create a new event name.

**Comment by glepnir** [comment]
right ..forget that.

**Comment by clason** [comment]
> Should be a User event. You cannot create a new event name.

Specifically, the event would be `User` with `pattern = 'LanguageType'`.

**Comment by glepnir** [comment]
now looks right

**Comment by stevenproctor** [comment]
👍 ♥️

**Comment by stevenproctor** [comment]
Please forgive my hopefully last set of related questions in the hope to understand the bigger picture.

Looking at the documentation, it does not appear there is any ability to pass extra data to the `User` autocommand other than the `pattern`.

This means that my code that might listen to the various other autocommands, like `CursorMoved`, `CursorHold`, etc., would look at the current node, determine the language (and maybe if it has changed), and then fire off the event.

The "subscriber" of the autocommand would also be expected to do the same to get the node at the current position, determine its language, and then do their updates based off the language.

I assume that this is because of the possible frequent and asynchronous nature of these events that this is how User commands are intended to behave.

Am I understanding this correctly?

**Comment by clason** [comment]
Yes, you can pass arbitrary data, see `:h nvim_exec_autocmds`.

In any case, this is -- if you forgive the formulation -- starting to sound like an XY problem: You're trying to solve a problem X via approach Y, so you ask about Y even though that's not the best way of tackling X. It'd be more appropriate to discuss this on Matrix.
